CI: add Runtime DROP_ALL test

manalibhutiyani opened the pull request:
This test adds the runtime `DROP_ALL`  tests and does 3 checks to make sure DROP_ALL is applied properly

1. It starts the container with no labels and checks if state is waiting-for-identity
2. Checks if the corresponding lxc_config.h contains the DROP_ALL macro
3. Checks if cilium monitor shows the correct L3 deny message for the endpoint.

For ingress, we try to ping from host to the container without labels.
For egress, we create a container and make it execute ping to 8.8.8.8

Fixes: #3625
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

ianvernon commented on test/runtime/Policies.go:
```diff
@@ -1393,3 +1394,101 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
+
+	BeforeEach(func() {
+		once.Do(initialize)
+		vm.PolicyDelAll()
```
Do not ignore the return value of this function. Do the following:
```
res := vm.PolicyDelAll()
res.ExpectSuccess("waiting for all policies to be deleted failed")
```

ianvernon commented on test/runtime/Policies.go:
```diff
@@ -1393,3 +1394,101 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
+		res.ExpectSuccess("Setting policyEnforcement to default")
+	})
+
+	AfterEach(func() {
```
Change this to the following:
```
JustAfterEach(func() {
	vm.ValidateNoErrorsOnLogs(CurrentGinkgoTestDescription().Duration)
})

AfterFailed(func() {
	vm.ReportFailed()
})
```

and also in `AfterEach`, delete all policies : 

```
res := vm.PolicyDelAll()
res.ExpectSuccess("waiting for all policies to be deleted failed")
```

manalibhutiyani replied:
There are no policies used in the test.

ianvernon replied:
> There are no policies used in the test.

Right you are, my bad :)

ianvernon commented on test/runtime/Policies.go:
```diff
@@ -1393,3 +1394,101 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
+		cmdStr := "docker run -dt --name ingressContainer --net=cilium-net tgraf/netperf"
+		vm.Exec(cmdStr).ExpectSuccess("Failed to create docker container")
+
+		endpointCmd := "cilium endpoint list | grep \"waiting-for-identity\" | awk '{ print $1}'"
```
Avoid using `awk` and `grep`; it's brittle and is prone to break whenever the CLI output changes. Instead, do something like the following:

```
ingressContainer := "ingressContainer"
...
endpointIDS, err := vm.GetEndpointsIds()
Expect(err).To(BeNil(), "Unable to get IDs of endpoints")

ingressEpID, exists := endpointIDS[ingressContainer]
Expect(exists).To(BeTrue(), "Expected endpoint ID to exist for %s", ingressContainer)

ingressEpModel := vm.EndpointGet(ingressEpID)
Expect(ingressEpModel).To(Not(BeNil()), "nil model returned for endpoint %s", ingressEpID)
```

Then, you can access the field from the endpoint model directly. This will keep the test stable since the API has been declared stable. Do the same on line 1438.

manalibhutiyani replied:
sure..will look into it

ianvernon commented on test/runtime/Policies.go:
```diff
@@ -1393,3 +1394,101 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
+		err := res.WaitUntilMatch("drop (Policy denied (L3))")
+		Expect(err).To(BeNil(), "DROP all on egress failed.")
+
+		vm.ContainerRm("ingressContainer").ExpectSuccess("Container server cannot be deleted")
```
Do not remove containers as part of the `It` itself. Resources should be removed in the `AfterEach` construct. This is due to the following:

1. We always want to make unconditionally that resources are cleaned up after each test run. `AfterEach` is ran whether a test succeeds or fails; an `It` exits if any `Expect` statement fails, and does not run the rest of the statements in the `It`. Since for builds of master, we run all tests even if one of them fails, this means that if the code is kept as it is in its current form, there will be leftover cruft which could potentially affect the state of other tests. Since there is another `It` which creates a container, I'd recommend wrapping each `It` within a `Context`. I can give you more information about this offline, as this is easier to explain in person.
2. If the container is removed in the `It`, then if logs are gathered, then the container will already be removed and we will have no information from the Cilium CLI for the container as it won't exist any more.

manalibhutiyani replied:
There is a reason for not adding the container creation in the BeforeEach function. The way the container is created is different i.e we need to create a docker container without any labels and also for egress we need to create a container without labels and start ping immediately on it. Hence this code cannot be common in `BeforeEach`

ianvernon replied:
@manalibhutiyani this is why I said the following

> Since there is another It which creates a container, I'd recommend wrapping each It within a Context. I can give you more information about this offline, as this is easier to explain in person

Let's discuss offline.

ianvernon commented on test/runtime/Policies.go:
```diff
@@ -1393,3 +1394,101 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
+
+		By("Testing cilium monitor drop")
+		err := res.WaitUntilMatch("drop (Policy denied (L3))")
+		Expect(err).To(BeNil(), "DROP all on egress failed.")
```
egress --> ingress?

ianvernon commented on test/runtime/Policies.go:
```diff
@@ -1393,3 +1394,101 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
+
+		By("Testing ingress with ping from localhost to endpoint with no labels")
+		lxcCmd = fmt.Sprintf("ping %s", endpointIP)
+		vm.Exec(lxcCmd)
```
What is the reason for not checking the result of this command?

Use `test/helpers/wrappers.go:Ping()` for the command. Then, use 
```
res := vm.Exec(lxcmd)
res.ExpectFail("Unexpectedly able to ping endpoint with DROP_ALL ingress policy")
```

Then after this you can check the output of the monitor for `Policy denied (L3)`?

ianvernon commented on test/runtime/Policies.go:
```diff
@@ -1393,3 +1394,101 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
+		cmdStr := "docker run -dt --name egressContainer --net=cilium-net tgraf/netperf ping 8.8.8.8"
+		vm.Exec(cmdStr).ExpectSuccess("Failed to create docker container")
+
+		endpointCmd := "cilium endpoint list | grep \"waiting-for-identity\" | awk '{ print $1}'"
```
See above comment for not using `awk` and `grep`.

ianvernon commented on test/runtime/Policies.go:
```diff
@@ -1393,3 +1394,101 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
+		err := res.WaitUntilMatch("drop (Policy denied (L3))")
+		Expect(err).To(BeNil(), "DROP all on egress failed.")
+
+		vm.ContainerRm("egressContainer").ExpectSuccess("Container server cannot be deleted")
```
Do not remove resources in `It`. See comment above.

manalibhutiyani replied:
Since I created the container in the `It` function, I deleted it in the same function. I can move the deletion in the `AfterEach`, but I don't this I can make the creation generic in `BeforeEach`

ianvernon reviewed (changes requested):
Thank you for adding this test! Please see my comments.

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1393,3 +1394,101 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
 	})
 })
+
+var _ = Describe("RuntimeValidatedPolicyDropAllTests", func() {
```
Please, add the test in `RuntimeValidatedPolicies` testgroup, doesn't make sense to me to use another test-group to do the same.

manalibhutiyani replied:
`RuntimeValidatedPolicies` creates a bunch of containers in the `BeforeAll` which are not needed for this test. Please read the container creation for this test carefully. `SampleContainersActions` creates containers with labels by default. Also please note that the container creation for` ingress` and `egress` is different for this test. Hence I did not see a point of coupling this with `RuntimeValidatedPolicies`. It should be a different test group.

eloycoto replied:
@manalibhutiyani the other containers can be in there without problems. You create a new container and you make the test based on that container. You should create a new `Context(`

manalibhutiyani replied:
@eloycoto The problem is 
```
BeforeAll(func() {
		****
		areEndpointsReady := vm.WaitEndpointsReady()
		Expect(areEndpointsReady).Should(BeTrue(), "Endpoints are not ready after timeout")
	})
```
`vm.WaitEndpointsReady` will hang because in this case, we are always in `waiting-for-identity` state.


I have tried this out and this will hang for the DROP_ALL tests. Also, we are unnecessary creating containers which this test does not need.  I think we do not need to overkill the whole test-grouping process  and group tests logically. Please note that drop_all is not a policy based test.

manalibhutiyani replied:
@eloycoto : This is what I get if I try to incorporate this as part of `RuntimeValidatedPolicies`

```
   /Users/manalibhutiyani/go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:312

    Unable to delete all policies
    Expected
        <bool>: false
    to be true

    /Users/manalibhutiyani/go/src/github.com/cilium/cilium/test/runtime/Policies.go:441
------------------------------
RuntimeValidatedPolicies DROP_ALL Egress Policy test 
  DROP_ALL Egress Policy test

```

this is being hit in 
```
AfterEach(func() {
		vm.PolicyDelAll().ExpectSuccess("Unable to delete all policies")
	})
```
Since DROP_ALL test do not have any policies, this global `AfterEach` simply fails.

ianvernon replied:
I am in agreement with Manali here. The purpose of `Describe` and `Context` is for logical organization of tests. Given that these tests use an entirely different topology for testing (containers with no labels vs. the containers with labels in `RuntimeValidatedPolicies`), it doesn't make sense to group them under the same `Describe`. 

Ginkgo's own documentation says: ```You should make use of Describe and Context containers to expressively organize the behavior of your code.```

Given that we aren't really testing policy (we are testing behavior with lack of identity and how the datapath is configured), this further emphasizes the need for a separate `Describe`. I am for adhering to a sense of organization for tests, but we need to make sure that it is easy for people to add tests and not focus on the minutiae of how things are structured. There are some exceptions to this in my opinion (making sure that resources are created / deleted in the right places to gather logs, making sure constants are used instead of hardcoding variables, proper documentation, etc.), but I think that we can always refactor as time goes on. This PR adds significant value in testing a critical part of how our datapath functions, especially from a security standpoint. I'm for keeping @manalibhutiyani 's addition within its own `Describe`, and we can discuss post-merge if we really want to do it the other way.

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1393,3 +1394,101 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
+	var vm *helpers.SSHMeta
+	var dropAll bool
+
+	initialize := func() {
```
This will not  be used if you move to `RuntimeValidatedPolicies` but in this case, you should use `BeforeAll`

manalibhutiyani replied:
There is a reason for not adding the container creation in the BeforeEach function. The way the container is created is different i.e we need to create a docker container without any labels and also for egress we need to create a container without labels and start ping immediately on it. Hence this code cannot be common in `BeforeEach`

eloycoto replied:
This doesn't make sense to me, you can create a Context in the test and do the things that you have done in the test in there.

manalibhutiyani replied:
maybe there is a miscommunication here. I will discuss this with @ianvernon and get back

manalibhutiyani replied:
Also if I make this part of `RuntimeValidatedPolicies`, 
```
         AfterAll(func() {
		vm.PolicyDelAll().ExpectSuccess("Unable to delete all policies")
		vm.SampleContainersActions(helpers.Delete, helpers.CiliumDockerNetwork)
	})
``` 
Since there is no policy in this test, this will hang altogether. I will be following the `context` review suggestion made by @ianvernon but will be keeping the separate `testGroup`.

eloycoto replied:
Still does not make sense to me. The `AfterAll` is a global method that we make sure that all is clean when it finish. That will not fail anyway. Test can be in a RuntimePolicies tests without having policies in multiple test groups and duplicated code.

manalibhutiyani replied:
I pointed out the wrong. function:

```
BeforeAll(func() {
		****
		areEndpointsReady := vm.WaitEndpointsReady()
		Expect(areEndpointsReady).Should(BeTrue(), "Endpoints are not ready after timeout")
	})
```

I have tried this out and this will hang for the DROP_ALL tests. Also, we are unnecessary creating containers which this test does not need. I think we do not need to overkill the whole test-grouping process and group tests logically. Please note that drop_all is not a policy based test.

manalibhutiyani replied:
@eloycoto : This is what I get if I try to incorporate this as part of `RuntimeValidatedPolicies`

```
   /Users/manalibhutiyani/go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:312

    Unable to delete all policies
    Expected
        <bool>: false
    to be true

    /Users/manalibhutiyani/go/src/github.com/cilium/cilium/test/runtime/Policies.go:441
------------------------------
RuntimeValidatedPolicies DROP_ALL Egress Policy test 
  DROP_ALL Egress Policy test

```

this is being hit in 
```
AfterEach(func() {
		vm.PolicyDelAll().ExpectSuccess("Unable to delete all policies")
	})
```
Since DROP_ALL test do not have any policies, this global `AfterEach` simply fails.

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1393,3 +1394,101 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
+		logger.Info("Starting")
+		vm = helpers.CreateNewRuntimeHelper(helpers.Runtime, logger)
+		areEndpointsReady := vm.WaitEndpointsReady()
+		Expect(areEndpointsReady).Should(BeTrue())
```
please, use assert messages.

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1393,3 +1394,101 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
+		defer cancel()
+
+		By("Create an endpoint with no labels to test DROP_ALL")
+		cmdStr := "docker run -dt --name ingressContainer --net=cilium-net tgraf/netperf"
```
`vm.ContainerCreate()` instead of hardcode

manalibhutiyani replied:
There is a reason for not adding the container creation in the BeforeEach function. The way the container is created is different i.e we need to create a docker container without any labels and also for egress we need to create a container without labels and start ping immediately on it. Hence this code cannot be common in `BeforeEach`

manalibhutiyani replied:
I guess I can use `vm.ContainerCreate()` with different options. Let me try that.

eloycoto replied:
but you still can use the vm.ContainerCreate() no?

manalibhutiyani replied:
Yes, I think I can do it with `vm.ContainerCreate()`

manalibhutiyani replied:
Looks like `vm.ContainerCreate` does not handle options to execute on container startup time correctly. 
```
res := vm.ContainerCreate(dropAllContainer, "tgraf/netperf", helpers.CiliumDockerNetwork, "ping 8.8.8.8")
```
translates to `docker run -dt --name dropAllContainer --net=cilium-net ping 8.8.8.8 tgraf/netperf `

which is incorrect 
```
docker run -dt --name dropAllContainer --net=cilium-net ping 8.8.8.8 tgraf/netperf
Unable to find image 'ping:latest' locally
docker: Error response from daemon: pull access denied for ping, repository does not exist or may require 'docker login'.
```

manalibhutiyani replied:
We need to append the command to execute on container start after the image in `vm.ContainerCreate`

Something like this
```
// ContainerCreate is a wrapper for `docker run`. It runs an instance of the
// specified Docker image with the provided network, name, and options.
func (s *SSHMeta) ContainerCreate(name, image, net, options string, cmdOnStart string) *CmdRes {
	cmd := fmt.Sprintf(
		"docker run -d --name %s --net %s %s %s %s", name, net, options, image, cmdOnStart)
	log.Debugf("spinning up container with command '%v'", cmd)
	return s.ExecWithSudo(cmd)
}
```
@ianvernon @eloycoto

manalibhutiyani replied:
done

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1393,3 +1394,101 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
+
+		By("Testing endpoint lxc_config.h for DROP_ALL")
+
+		lxcCmd := fmt.Sprintf("sudo cat /var/run/cilium/state/%s/lxc_config.h | grep 'DROP_ALL'", endpointID)
```
don't use sudo here, and run the lxcCmd with `vm.ExecWithSudo` please

manalibhutiyani replied:
I was getting an error/hang with that function and resorted to this. I will try to use it and see if it works, but if it does not, I can do it as a separate PR and debug it later, but I don't think I am debugging it as part of this PR.

eloycoto replied:
Please ping me if you see errors with `ExecWithSudo()`

manalibhutiyani replied:
done

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1393,3 +1394,101 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
+		lxcCmd := fmt.Sprintf("sudo cat /var/run/cilium/state/%s/lxc_config.h | grep 'DROP_ALL'", endpointID)
+
+		dropAll = false
+		if strings.TrimSpace(vm.Exec(lxcCmd).GetStdOut()) == "#define DROP_ALL" {
```
```
res = vm.ExecWithSudo(lxcCmd)
res.ExpectSucess("Cannot get lxc_config")
res.ExpectContains("#define DROP_ALL", "DROP_ALL is not defined")
```

manalibhutiyani replied:
sure

manalibhutiyani replied:
done

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1393,3 +1394,101 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
+
+		By("Testing endpoint lxc_config.h for DROP_ALL")
+
+		lxcCmd := fmt.Sprintf("sudo cat /var/run/cilium/state/%s/lxc_config.h | grep 'DROP_ALL'", endpointID)
```
same comment as above.

manalibhutiyani replied:
done

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1393,3 +1394,101 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
+		lxcCmd := fmt.Sprintf("sudo cat /var/run/cilium/state/%s/lxc_config.h | grep 'DROP_ALL'", endpointID)
+
+		dropAll = false
+		if strings.TrimSpace(vm.Exec(lxcCmd).GetStdOut()) == "#define DROP_ALL" {
```
Here the `ExpectContains(` as above

manalibhutiyani replied:
done

joestringer commented on test/runtime/Policies.go:
```diff
@@ -1393,3 +1394,101 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
+		vm.Exec(lxcCmd)
+
+		By("Testing cilium monitor drop")
+		err := res.WaitUntilMatch("drop (Policy denied (L3))")
```
I wonder if we could reuse the actual string from the monitor package?

https://github.com/cilium/cilium/blob/0d745b8acf29139e88fe1fb7d4f9c0441082d158/pkg/monitor/datapath_drop.go#L47

(I'd like to rename it at some point, this would make it easier)

ianvernon replied:
@joestringer the only concern I have with that is that we want to avoid getting code involving bpf compilation into the Ginkgo tests because it requires compilation on the host running tests, and since Macs do not have the required header files, the tests may not compile if we do this. If this is the case, we could factor out these strings into a subpackage within monitor and then import?

joestringer replied:
`pkg/monitor` has nothing to do with BPF compilation. The only dependency I see that might be extraneous is the accesslog stuff, I don't know how deep that dependency tree might go. Either way, addressing this can be best-effort and should not block adding this test.

ianvernon replied:
@joestringer true. In the past though I have found surprising transitive dependencies ;)

Commit:
CI: add Runtime default DROP_ALL test
This test adds the runtime DROP_ALL tests and does 3 checks to make sure DROP_ALL is applied properly

1. It starts the container with no labels and checks if state is waiting-for-identity
2. Checks is the corresponding lxc_config.h contains the DROP_ALL macro
3. Checks if cilium monitor shows the correct L3 deny message for the endpoint.

For ingress, we try to ping from host to the container without labels.
For egress, we create a container and make it execute ping to 8.8.8.8

It also extends the helper functions:
1. vm.ContainerCreate is extended to include cmd to run immediately
after starting the container
2. GetAllEndpointsIds is added to return all endpoints without filtering out and endpoints
   This was needed to filter endpoints with no labels.

Finally, this commit wraps
the ingress DROP_ALL and egress DROP_ALL in separate ginkgo
contexts to allow creation/deletion of label-less containers
differently.

Fixes: #3625
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

eloycoto commented on test/helpers/cilium.go:
```diff
@@ -309,6 +309,19 @@ func (s *SSHMeta) GetEndpointsIDMap() (map[string]string, error) {
+// GetAllEndpointsIds returns a mapping of all Docker container name to to its
+// corresponding endpoint ID, and an error if the list of endpoints cannot be
+// retrieved via the Cilium CLI.
+func (s *SSHMeta) GetAllEndpointsIds() (map[string]string, error) {
```
I'm not sure why you add this function instead of using `GetEndpointsIds` could you explain why?

manalibhutiyani replied:
filter := `{range [?(@.status.labels.security-relevant[0]!="reserved:health")]}{@.status.external-identifiers.container-name}{"="}{@.id}{"\n"}{end}`

Since we do not have labels, this `status.labels.security-relevant[0]` does not work.

I will not be changing this as part of this PR.

eloycoto replied:
Make sense. Thanks

eloycoto commented on test/helpers/docker.go:
```diff
@@ -32,9 +32,10 @@ func (s *SSHMeta) ContainerExec(name string, cmd string, optionalArgs ...string)
 // ContainerCreate is a wrapper for `docker run`. It runs an instance of the
 // specified Docker image with the provided network, name, and options.
-func (s *SSHMeta) ContainerCreate(name, image, net, options string) *CmdRes {
+func (s *SSHMeta) ContainerCreate(name, image, net, options string, cmdOnStart string) *CmdRes {
```
Could you move this cmdOnStart as variadic argument? It's only used one time and you had changed how the function works, I'll do something like this: 
```
func (s *SSHMeta) ContainerCreate(name, image, net, options string, cmdParams ...string) *CmdRes {
	cmdOnStart := ""
	if len(cmdParams) > 0 {
		cmdOnStart = strings.Join(cmdParams, " ")
	}

	cmd := fmt.Sprintf(
		"docker run -d --name %s --net %s %s %s %s", name, net, options, image, cmdOnStart)
	log.Debugf("spinning up container with command '%v'", cmd)
	return s.ExecWithSudo(cmd)
}
```

This allow to keep other functions as it was, and no new argument at all. In the case that you need to run a specific command can be run with the variadic argument.

manalibhutiyani replied:
sure

manalibhutiyani replied:
done

eloycoto commented on test/helpers/docker.go:
```diff
@@ -32,9 +32,10 @@ func (s *SSHMeta) ContainerExec(name string, cmd string, optionalArgs ...string)
 
 // ContainerCreate is a wrapper for `docker run`. It runs an instance of the
 // specified Docker image with the provided network, name, and options.
```
If you add a new argument to the function the comment should be updated.

manalibhutiyani replied:
done

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1478,3 +1479,125 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
+			defer cancel()
+
+			endpoints, err := vm.GetAllEndpointsIds()
+			Expect(err).Should(BeNil())
```
assert message

manalibhutiyani replied:
Please give me an example. `BeNil()` is an assert and you have used this throughout Policies.go. In fact this is simply a copy of the `RuntimeValidatedPolicies`
See `It("L3/L4 Checks", func() {` . 

I would like to focus on the main test first and not nits at this point. Thanks!

eloycoto replied:
@manalibhutiyani yes, there is a lot of places that there is no AssertMessages, it has been a rule that in the new code we added the assert message.

manalibhutiyani replied:
Done. Added assert message
```
Expect(err).Should(BeNil(), "Unable to get IDs of endpoints")
```

ianvernon commented on test/runtime/Policies.go:
```diff
@@ -1478,3 +1479,125 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
+	}
+
+	BeforeAll(func() {
+		once.Do(initialize)
```
Nit: (non blocking) - you can just merge the code within `initialize` in this function and get rid of `initialize` because `BeforeAll` is ran only once, so having the code wrapped in a `sync.Once` is redundant.

manalibhutiyani replied:
done

ianvernon commented on test/runtime/kafka.go:
```diff
@@ -58,13 +58,13 @@ var _ = Describe("RuntimeValidatedKafka", func() {
 
 			vm.ContainerCreate("kafka", "wurstmeister/kafka", helpers.CiliumDockerNetwork, fmt.Sprintf(
-				"-l id.kafka -e KAFKA_ZOOKEEPER_CONNECT=%s:2181 -e KAFKA_ZOOKEEPER_SESSION_TIMEOUT_MS=20000 -e KAFKA_LISTENERS=PLAINTEXT://:9092 -e KAFKA_ZOOKEEPER_CONNECTION_TIMEOUT_MS=20000", zook["IPv4"]))
+				"-l id.kafka -e KAFKA_ZOOKEEPER_CONNECT=%s:2181 -e KAFKA_ZOOKEEPER_SESSION_TIMEOUT_MS=20000 -e KAFKA_ZOOKEEPER_CONNECTION_TIMEOUT_MS=20000", zook["IPv4"]), "")
```
Can you explain why ` KAFKA_LISTENERS=PLAINTEXT://:9092 -e ` is removed here?

manalibhutiyani replied:
This looks like a rebase issue, I will undo it. Thanks for catching!

manalibhutiyani replied:
done

eloycoto commented on test/helpers/docker.go:
```diff
@@ -31,10 +31,16 @@ func (s *SSHMeta) ContainerExec(name string, cmd string, optionalArgs ...string)
 	cmd := fmt.Sprintf(
-		"docker run -d --name %s --net %s %s %s", name, net, options, image)
+		"docker run -d --name %s --net %s %s %s %s", name, net, options, image, cmdOnStart)
+	log.Debugf("spinning up container with command '%v'", cmd)
```
This is not needed, `ExecWithSudo` will log the command.
